### PR TITLE
Shoppingcart practies

### PR DESCRIPTION
**Pull request by opopovych**

### CODE REVIEW

File: src/main/java/mate/academy/dto/cartitem/AddItemToCartRequestDto.java [hunk 1]
```diff
@@ -0,0 +1,10 @@
+package mate.academy.dto.cartitem;
+
+import jakarta.validation.constraints.Positive;
+
+public record AddItemToCartRequestDto(
+        Long bookId,
```
Comment (Rommelua) [inline src/main/java/mate/academy/dto/cartitem/AddItemToCartRequestDto.java:8]: NotNull

File: src/main/java/mate/academy/dto/cartitem/AddItemToCartRequestDto.java [hunk 2]
```diff
@@ -0,0 +1,10 @@
...
+        @Positive
+        int quantity
```
Comment (Rommelua) [inline src/main/java/mate/academy/dto/cartitem/AddItemToCartRequestDto.java:11]: NotNull

File: src/main/java/mate/academy/model/ShoppingCart.java [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+import jakarta.persistence.JoinColumn;
+import jakarta.persistence.OneToMany;
+import jakarta.persistence.OneToOne;
+import jakarta.persistence.Table;
+import java.util.Set;
+import lombok.Data;
+import lombok.NoArgsConstructor;
+
+@Entity
+@Data
```
Comment (Rommelua) [inline src/main/java/mate/academy/model/ShoppingCart.java]: do not use `@Data` with Entity classes

File: src/main/java/mate/academy/model/CartItem.java [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+import jakarta.persistence.Id;
+import jakarta.persistence.JoinColumn;
+import jakarta.persistence.ManyToOne;
+import jakarta.persistence.Table;
+import lombok.Data;
+
+@Entity
+@Data
+@Table(name = "cart_items")
+public class CartItem {
```
Comment (Rommelua) [inline src/main/java/mate/academy/model/CartItem.java:21]: image

File: src/main/java/mate/academy/repository/cartitem/CartItemRepository.java [hunk 5]
```diff
@@ -0,0 +1,12 @@
...
+
+import mate.academy.model.CartItem;
+import org.springframework.data.jpa.repository.JpaRepository;
+import org.springframework.data.jpa.repository.Modifying;
+import org.springframework.data.jpa.repository.Query;
+
+public interface CartItemRepository extends JpaRepository<CartItem, Long> {
+    @Modifying
+    @Query("DELETE FROM CartItem c WHERE c.shoppingCart.id = :cartId")
+    void deleteByShoppingCartId(Long cartId);
```
Comment (Rommelua) [inline src/main/java/mate/academy/repository/cartitem/CartItemRepository.java:13]: image

File: src/main/java/mate/academy/service/ShoppingCartService.java [hunk 6]
```diff
@@ -0,0 +1,24 @@
+package mate.academy.service;
+
+import mate.academy.dto.cartitem.AddItemToCartRequestDto;
+import mate.academy.dto.cartitem.UpdateCartItemRequestDto;
+import mate.academy.dto.shoppingcart.ShoppingCartDto;
+import mate.academy.model.ShoppingCart;
+import mate.academy.model.User;
+
+public interface ShoppingCartService {
+    ShoppingCart createShoppingCart(User user);
```
Comment (Rommelua) [inline src/main/java/mate/academy/service/ShoppingCartService.java]: do we use it?

File: src/main/java/mate/academy/service/ShoppingCartService.java [hunk 7]
```diff
@@ -0,0 +1,24 @@
...
+
+    ShoppingCartDto addBooksToCartByUserId(Long userId, AddItemToCartRequestDto requestDto);
```
Comment (Rommelua) [inline src/main/java/mate/academy/service/ShoppingCartService.java]:
```suggestion
    ShoppingCartDto addBookToCartByUserId(Long userId, AddItemToCartRequestDto requestDto);
```

File: src/main/java/mate/academy/service/ShoppingCartService.java [hunk 8]
```diff
@@ -0,0 +1,24 @@
...
+
+    ShoppingCartDto getCartDtoByUserId(Long id);
```
Comment (Rommelua) [inline src/main/java/mate/academy/service/ShoppingCartService.java]:
```suggestion
    ShoppingCartDto getCartByUserId(Long id);
```

File: src/main/java/mate/academy/service/ShoppingCartService.java [hunk 9]
```diff
@@ -0,0 +1,24 @@
...
+
+    ShoppingCart getCartByUserId(Long id);
```
Comment (Rommelua) [inline src/main/java/mate/academy/service/ShoppingCartService.java]:
```suggestion
```

File: src/main/java/mate/academy/service/impl/ShoppingCartServiceImpl.java [hunk 10]
```diff
@@ -0,0 +1,92 @@
...
+    @Override
+    public ShoppingCart createShoppingCart(User user) {
+        ShoppingCart shoppingCart = new ShoppingCart();
+        shoppingCart.setUser(user);
+        shoppingCart.setCartItems(new HashSet<>());
+        return shoppingCartRepository.save(shoppingCart);
+    }
+
+    @Override
+    public ShoppingCartDto addBooksToCartByUserId(Long userId, AddItemToCartRequestDto requestDto) {
```
Comment (Rommelua) [inline src/main/java/mate/academy/service/impl/ShoppingCartServiceImpl.java]: Add `@Transactional`

File: src/main/java/mate/academy/service/impl/ShoppingCartServiceImpl.java [hunk 11]
```diff
@@ -0,0 +1,92 @@
...
+        ShoppingCart cartFromDb = getCartFromDb(userId);
+        CartItem requestedItem = cartItemMapper.toModel(requestDto);
+        requestedItem.setShoppingCart(cartFromDb);
+        cartItemRepository.save(requestedItem);
+        return cartMapper.toDto(getCartFromDb(userId));
```
Comment (Rommelua) [inline src/main/java/mate/academy/service/impl/ShoppingCartServiceImpl.java]:
```suggestion
        return cartMapper.toDto(cartFromDb);
```

**Review by Rommelua — CHANGES_REQUESTED** [review]

**Comment by opopovych** [comment]
i can not understand, why all checks was failed. i think, that i did all correctly.

File: src/main/java/mate/academy/repository/cartitem/CartItemRepository.java (hunk 5 above)
Comment (Rommelua) [reply src/main/java/mate/academy/repository/cartitem/CartItemRepository.java:13]: not fixed - use soft delete

**Review by Rommelua — CHANGES_REQUESTED** [review]

File: src/main/java/mate/academy/controller/ShoppingCartController.java [hunk 12]
```diff
@@ -0,0 +1,69 @@
...
+import org.springframework.web.bind.annotation.PathVariable;
+import org.springframework.web.bind.annotation.PostMapping;
+import org.springframework.web.bind.annotation.PutMapping;
+import org.springframework.web.bind.annotation.RequestBody;
+import org.springframework.web.bind.annotation.RequestMapping;
+import org.springframework.web.bind.annotation.RestController;
+
+@Tag(name = "Shopping cart manager", description = "Endpoints for managing shopping carts")
+@RestController
+@RequestMapping("/api/cart")
```
Comment (okuzan) [inline src/main/java/mate/academy/controller/ShoppingCartController.java]: extract api to server servlet context-path property

File: src/main/java/mate/academy/controller/ShoppingCartController.java [hunk 13]
```diff
@@ -0,0 +1,69 @@
...
+public class ShoppingCartController {
+    private final ShoppingCartService shoppingCartService;
+
+    @PreAuthorize("hasAuthority('USER')")
+    @Operation(summary = "Get the shopping cart",
+            description = "Get user's shopping cart, show what is in it")
+    @GetMapping
+    ShoppingCartDto getCart(Authentication authentication) {
+        User user = (User) authentication.getPrincipal();
+        return shoppingCartService.getCartByUserId(user.getId());
```
Comment (okuzan) [inline src/main/java/mate/academy/controller/ShoppingCartController.java]: maybe let's create a private method to get userId from authentication?

File: src/main/java/mate/academy/model/CartItem.java [hunk 14]
```diff
@@ -0,0 +1,37 @@
...
+import lombok.Getter;
+import lombok.Setter;
+import org.hibernate.annotations.SQLDelete;
+import org.hibernate.annotations.Where;
+
+@Entity
+@Getter
+@Setter
+@SQLDelete(sql = "UPDATE cart_items SET deleted = true WHERE id=?")
+@Where(clause = "deleted=false")
```
Comment (okuzan) [inline src/main/java/mate/academy/model/CartItem.java]:
```suggestion
@SQLDelete(sql = "UPDATE cart_items SET deleted = TRUE WHERE id = ?")
@SQLRestriction(clause = "deleted = FALSE")
```

File: src/main/java/mate/academy/model/ShoppingCart.java [hunk 15]
```diff
@@ -0,0 +1,34 @@
...
+import lombok.NoArgsConstructor;
+import lombok.Setter;
+
+@Entity
+@Getter
+@Setter
+@NoArgsConstructor
+@Table(name = "shopping_carts")
+public class ShoppingCart {
+
```
Comment (okuzan) [inline src/main/java/mate/academy/model/ShoppingCart.java]:
```suggestion
```

File: src/main/java/mate/academy/model/CartItem.java [hunk 16]
```diff
@@ -0,0 +1,37 @@
...
+@Table(name = "cart_items")
+public class CartItem {
+
```
Comment (okuzan) [inline src/main/java/mate/academy/model/CartItem.java]:
```suggestion
```

File: src/main/java/mate/academy/service/ShoppingCartService.java [hunk 17]
```diff
@@ -0,0 +1,19 @@
+package mate.academy.service;
+
+import mate.academy.dto.cartitem.AddItemToCartRequestDto;
+import mate.academy.dto.cartitem.UpdateCartItemRequestDto;
+import mate.academy.dto.shoppingcart.ShoppingCartDto;
+
+public interface ShoppingCartService {
+
```
Comment (okuzan) [inline src/main/java/mate/academy/service/ShoppingCartService.java:8]:
```suggestion
```

File: src/main/resources/db/changelog/changes/10-create-shopping-cart-table.yaml [hunk 18]
```diff
@@ -0,0 +1,25 @@
...
+              - column:
+                  name: user_id
+                  type: BIGINT
+                  constraints:
+                    nullable: false
+                    foreignKeyName: fk_shopping_carts_user_id
+                    references: users(id)
+      rollback:
+        - dropTable:
+            tableName: shopping_carts
```
Comment (okuzan) [inline src/main/resources/db/changelog/changes/10-create-shopping-cart-table.yaml]:
```suggestion
            tableName: shopping_carts

```
add empty line wherever you see this red symbol in your PR

Comment (okuzan) [reply src/main/resources/db/changelog/changes/10-create-shopping-cart-table.yaml]: image

**Review by okuzan — CHANGES_REQUESTED** [review]

File: src/main/java/mate/academy/model/CartItem.java [hunk 19]
```diff
@@ -0,0 +1,37 @@
...
+import lombok.Getter;
+import lombok.Setter;
+import org.hibernate.annotations.SQLDelete;
+import org.hibernate.annotations.SQLRestriction;
+
+@Entity
+@Getter
+@Setter
+@SQLDelete(sql = "UPDATE cart_items SET deleted = TRUE WHERE id = ?")
+//@Where(clause = "deleted = FALSE")
```
Comment (MentorMA) [inline src/main/java/mate/academy/model/CartItem.java]: Commented code

File: src/main/java/mate/academy/repository/cartitem/CartItemRepository.java [hunk 20]
```diff
@@ -0,0 +1,16 @@
+package mate.academy.repository.cartitem;
+
+import mate.academy.model.CartItem;
+import org.springframework.data.jpa.repository.JpaRepository;
+import org.springframework.data.jpa.repository.Modifying;
+import org.springframework.data.jpa.repository.Query;
+import org.springframework.stereotype.Repository;
+
+@Repository
```
Comment (MentorMA) [inline src/main/java/mate/academy/repository/cartitem/CartItemRepository.java]: This annotation is not mandatory

**Review by MentorMA — CHANGES_REQUESTED** [review]

**Pull request merged**